### PR TITLE
translate custom dataloader

### PR DESCRIPTION
**Pull request by MARD1NO**
翻译 自定义 dataloader

### CODE REVIEW

File: en/docs/extended_topics/implement_data_loader.md [hunk 1]
```diff
@@ -0,0 +1,461 @@
+# Customize DataLoader
+
+As described in [Data Input](../basics_topics/data_input.md), OneFlow supports two ways to load data: one is directly use Numpy data, the other one is use DataLoader and some relative operators. 
+
+Under the large industrial scene, Data Loading can easily become the bottleneck through the training process. Since we use DataLoader and some preprocessing operators, OneFlow's acceleration mechanism helps to load and preprocess data more efficiently, which can solve the problem.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:5]: Data -> data
大小写

Comment (doombeaker) [reply en/docs/extended_topics/implement_data_loader.md:5]: whicn can solve the problem. "the problem" 是上一句话的问题了，用"that problem"更好？

File: en/docs/extended_topics/implement_data_loader.md [hunk 2]
```diff
@@ -0,0 +1,461 @@
+# Customize DataLoader
+
+As described in [Data Input](../basics_topics/data_input.md), OneFlow supports two ways to load data: one is directly use Numpy data, the other one is use DataLoader and some relative operators. 
+
+Under the large industrial scene, Data Loading can easily become the bottleneck through the training process. Since we use DataLoader and some preprocessing operators, OneFlow's acceleration mechanism helps to load and preprocess data more efficiently, which can solve the problem. 
+
+To use DataLoader in OneFlow, we usually apply `XXXReader` to load the file data, and we use `XXXDecode` to decode or preprocess the data. These two operators work together to complete the function of DataLoader.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:7]: and we use `XXXDecode` -> and use `XXXDecode`

Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:7]: These two operators work together to complete the function of DataLoader.
->
These two operators work together to complete the function of data loading.

文章里把 Dataloader 当作一个专有名词，这个没什么问题（所以首字母大写）。但是因为是专有名词，所以这里不应该用来指代"数据加载“,而应该用 data loading 这种更通用的表述

File: en/docs/extended_topics/implement_data_loader.md [hunk 3]
```diff
@@ -0,0 +1,461 @@
+# Customize DataLoader
+
+As described in [Data Input](../basics_topics/data_input.md), OneFlow supports two ways to load data: one is directly use Numpy data, the other one is use DataLoader and some relative operators. 
+
+Under the large industrial scene, Data Loading can easily become the bottleneck through the training process. Since we use DataLoader and some preprocessing operators, OneFlow's acceleration mechanism helps to load and preprocess data more efficiently, which can solve the problem. 
+
+To use DataLoader in OneFlow, we usually apply `XXXReader` to load the file data, and we use `XXXDecode` to decode or preprocess the data. These two operators work together to complete the function of DataLoader.  
+
+Now OneFlow has built some data format's [DataLoader](https://oneflow.readthedocs.io/en/master/search.html?q=reader&check_keywords=yes&area=default) internally. If we want to use DataLoader to promote the efficiency of data loading, however, the DataLoader for the corresponding data format is not yet built in OneFlow. At this time, we can implement our own DataLoader to load the customized data format.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:9]: has built some data format's DataLoader -> has built some Dataloader

File: en/docs/extended_topics/implement_data_loader.md [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+
+Under the large industrial scene, Data Loading can easily become the bottleneck through the training process. Since we use DataLoader and some preprocessing operators, OneFlow's acceleration mechanism helps to load and preprocess data more efficiently, which can solve the problem. 
+
+To use DataLoader in OneFlow, we usually apply `XXXReader` to load the file data, and we use `XXXDecode` to decode or preprocess the data. These two operators work together to complete the function of DataLoader.  
+
+Now OneFlow has built some data format's [DataLoader](https://oneflow.readthedocs.io/en/master/search.html?q=reader&check_keywords=yes&area=default) internally. If we want to use DataLoader to promote the efficiency of data loading, however, the DataLoader for the corresponding data format is not yet built in OneFlow. At this time, we can implement our own DataLoader to load the customized data format.  
+
+In this article we implement a Mini Dataloader. You can check the [Complete Code](https://github.com/Oneflow-Inc/oneflow-documentation/tree/master/en/docs/code/extended_topics/data_loader/) in this repository.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:11]: Complete code -> code

File: en/docs/extended_topics/implement_data_loader.md [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+```
+
+This article will take Mini Dataloader as an example to explain the key points of implementing customized DataLoader. 
+
+# The composition of Dataloader
+
+A complete Dataloader generally includes two types of Op: 
+
+- Data Reader: Which is responsible for loading the data in file system to the input stream of memory and setting the data to the Op's output. 
+- Data Decoder: The Data Decoder decodes and output the data in Data Reader Op.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:34]: output -> outputs

File: en/docs/extended_topics/implement_data_loader.md [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+This article will take Mini Dataloader as an example to explain the key points of implementing customized DataLoader. 
+
+# The composition of Dataloader
+
+A complete Dataloader generally includes two types of Op: 
+
+- Data Reader: Which is responsible for loading the data in file system to the input stream of memory and setting the data to the Op's output. 
+- Data Decoder: The Data Decoder decodes and output the data in Data Reader Op. 
+
+For some simple data formats, which is no need for decoding. We can omit the Data Decoder and just use Data Reader.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:36]: no need for decoding. We  -> no need for decoding, we

File: en/docs/extended_topics/implement_data_loader.md [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+# The composition of Dataloader
+
+A complete Dataloader generally includes two types of Op: 
+
+- Data Reader: Which is responsible for loading the data in file system to the input stream of memory and setting the data to the Op's output. 
+- Data Decoder: The Data Decoder decodes and output the data in Data Reader Op. 
+
+For some simple data formats, which is no need for decoding. We can omit the Data Decoder and just use Data Reader. 
+
+As an example, since the data format processed by Mini Dataloader is simple. We still implement the Data Reader and Data Decoder. Among these two Ops:
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:38]: As an example, though the data format processed by Mini Dataloader is simple, we still implement the two types of ops: Data Reader and Data Decoder.

File: en/docs/extended_topics/implement_data_loader.md [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+A complete Dataloader generally includes two types of Op: 
+
+- Data Reader: Which is responsible for loading the data in file system to the input stream of memory and setting the data to the Op's output. 
+- Data Decoder: The Data Decoder decodes and output the data in Data Reader Op. 
+
+For some simple data formats, which is no need for decoding. We can omit the Data Decoder and just use Data Reader. 
+
+As an example, since the data format processed by Mini Dataloader is simple. We still implement the Data Reader and Data Decoder. Among these two Ops: 
+
+- `MiniReader` is responsible for reading data from files and split strings by commas. Convert the text to the float and set to the Op's output. The output shape is two columns of each row.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:40]: float  -> float value

File: en/docs/extended_topics/implement_data_loader.md [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+
+- Data Reader: Which is responsible for loading the data in file system to the input stream of memory and setting the data to the Op's output. 
+- Data Decoder: The Data Decoder decodes and output the data in Data Reader Op. 
+
+For some simple data formats, which is no need for decoding. We can omit the Data Decoder and just use Data Reader. 
+
+As an example, since the data format processed by Mini Dataloader is simple. We still implement the Data Reader and Data Decoder. Among these two Ops: 
+
+- `MiniReader` is responsible for reading data from files and split strings by commas. Convert the text to the float and set to the Op's output. The output shape is two columns of each row. 
+- `MiniDecoder` is responsible for splitting the two columns of each row output in above and get two outputs `x` and `y`, both of them shape is one column of each row.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:41]: both of their shape

File: en/docs/extended_topics/implement_data_loader.md [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+- Data Decoder: The Data Decoder decodes and output the data in Data Reader Op. 
+
+For some simple data formats, which is no need for decoding. We can omit the Data Decoder and just use Data Reader. 
+
+As an example, since the data format processed by Mini Dataloader is simple. We still implement the Data Reader and Data Decoder. Among these two Ops: 
+
+- `MiniReader` is responsible for reading data from files and split strings by commas. Convert the text to the float and set to the Op's output. The output shape is two columns of each row. 
+- `MiniDecoder` is responsible for splitting the two columns of each row output in above and get two outputs `x` and `y`, both of them shape is one column of each row. 
+
+In [test_mini_dataloader.py](https://github.com/Oneflow-Inc/oneflow-documentation/tree/master/en/docs/code/extended_topics/data_loader/test_mini_dataloader.py) we can see the use of both Ops at Python level:
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:43]: the use of -> the usage of

File: en/docs/extended_topics/implement_data_loader.md [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+
+We will introduce how to implement Data Reader and Data Decoder in C++ backend below. 
+
+# Data Reader operator
+
+## The class relationship in Data Reader
+
+We need to implement a class that inherits from `DataReader`, this class includes two important objects `loader_` and `parser_`, which inherits from `Dataset` and `Parser` separately. 
+
+- `loader_` 's job is to read data from the file system to buffer. The Op's author build the logic by overriding the `Next` function.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:68]: read data from -> load data from?

File: en/docs/extended_topics/implement_data_loader.md [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+## The class relationship in Data Reader
+
+We need to implement a class that inherits from `DataReader`, this class includes two important objects `loader_` and `parser_`, which inherits from `Dataset` and `Parser` separately. 
+
+- `loader_` 's job is to read data from the file system to buffer. The Op's author build the logic by overriding the `Next` function. 
+- `parser_` 's job is to set the data in buffer to Op's output. The Op's author build the logic by overriding the `Parser` function. 
+
+When Data Reader Op works, it will call the relative function in `loader_` to open files in the file system, and then call the `Next` function in `loader_` to read data from the file system according to the logic built by Op's author. 
+
+The pseudocode below shows the class relationship and the calling procedure. The actual code is more complicated than the pseudocode and it's not a one-to-one relationship:
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:73]: The actual code is more complicated and it is not the exact corresponding relationship.

File: en/docs/extended_topics/implement_data_loader.md [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+  //...
+};
+
+REGISTER_USER_KERNEL("MiniReader")
+    .SetCreateFn<MiniReaderKernel>()
+    .SetIsMatchedHob((user_op::HobDeviceTag() == "cpu")
+                     & (user_op::HobDataType("out", 0) == DataType::kDouble));
+```
+
+According to the knowledge in [Customize Op](../extended_topics/user_op.md), we have known that `MiniReaderKernel::Compute` function is responsible for the Op's compute logic. However, we use an override version of `Compute` that includes two parameters here. It's necessary to introduce the second parameter `OpKernelState`.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:170]: `MiniReaderKernel::Compute` function ->  `MiniReaderKernel::Compute` method
用method更准确点

File: en/docs/extended_topics/implement_data_loader.md [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+
+REGISTER_USER_KERNEL("MiniReader")
+    .SetCreateFn<MiniReaderKernel>()
+    .SetIsMatchedHob((user_op::HobDeviceTag() == "cpu")
+                     & (user_op::HobDataType("out", 0) == DataType::kDouble));
+```
+
+According to the knowledge in [Customize Op](../extended_topics/user_op.md), we have known that `MiniReaderKernel::Compute` function is responsible for the Op's compute logic. However, we use an override version of `Compute` that includes two parameters here. It's necessary to introduce the second parameter `OpKernelState`. 
+
+When we call the `Compute`, we need to maintain other objects in addition to get information from `ctx`. This type of object does not need to be created repeatedly, but their state of information may change as `Compute` function is called multiple times. In response to this need, OneFlow provides a override version with two parameters of `Compute`. In order to use it, we need to override `CreateOpKernelState` at the same time. `CreateOpKernelState` returns a `user_op::OpKernelState` derived class object. This object will be the second parameter to be transmitted when `Compute` is called.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:172]: This object will be the second parameter when `Compute` is called.

File: en/docs/extended_topics/implement_data_loader.md [hunk 15]
```diff
@@ -0,0 +1,461 @@
...
+```
+
+According to the knowledge in [Customize Op](../extended_topics/user_op.md), we have known that `MiniReaderKernel::Compute` function is responsible for the Op's compute logic. However, we use an override version of `Compute` that includes two parameters here. It's necessary to introduce the second parameter `OpKernelState`. 
+
+When we call the `Compute`, we need to maintain other objects in addition to get information from `ctx`. This type of object does not need to be created repeatedly, but their state of information may change as `Compute` function is called multiple times. In response to this need, OneFlow provides a override version with two parameters of `Compute`. In order to use it, we need to override `CreateOpKernelState` at the same time. `CreateOpKernelState` returns a `user_op::OpKernelState` derived class object. This object will be the second parameter to be transmitted when `Compute` is called. 
+
+So we only need to encapsulate the information we want to maintain, in addition to the `ctx`, as a derived class of `user_op::OpKernelState`, instantiate and return it in `CreateOpKernelState`.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:174]: encapsulate -> pack

File: en/docs/extended_topics/implement_data_loader.md [hunk 16]
```diff
@@ -0,0 +1,461 @@
...
+
+```c++
+  std::shared_ptr<user_op::OpKernelState>
+  CreateOpKernelState(user_op::KernelInitContext* ctx) override{
+    std::shared_ptr<MiniReaderWrapper> reader(new MiniReaderWrapper(ctx));
+    return reader;
+  }
+```
+
+In this way, OneFlow will call `CreateOpKernelState` function to create object automatically in appropriate time and transmit it to `Compute` as the second parameter. We can get this object in `Compute`, and use:
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:201]: transmit -> pass

Comment (doombeaker) [reply en/docs/extended_topics/implement_data_loader.md:201]: We can get this object in `Compute`, and use it:

File: en/docs/extended_topics/implement_data_loader.md [hunk 17]
```diff
@@ -0,0 +1,461 @@
...
+      loader_.reset(new RandomShuffleDataset<TensorBuffer>(ctx, std::move(loader_)));
+    }
+    int32_t batch_size = ctx->TensorDesc4ArgNameAndIndex("out", 0)->shape().elem_cnt();
+    loader_.reset(new BatchDataset<TensorBuffer>(batch_size, std::move(loader_)));
+    StartLoadThread();
+  }
+};
+```
+
+In addition to inheriting our `Dataset`'s `MiniDataset` class, OneFlow also build other `XXXDataset`, they can add additional features in the base of existing `Dataset`. For example, `RandomShuffleDataset` can be used to shuffle data, `BatchDataset` can be used to read batch data. When it is all done, we finally call `StartLoadThread`, which is used to start the loading thread. We will trigger the override function `MiniDataset::Next` in `StartLoadThread`.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:232]: 这里开始，三个反引号的格式解析错乱了，有编译看最终的效果吗?检查并留意下格式

Comment (MARD1NO) [reply en/docs/extended_topics/implement_data_loader.md:232]: image
结果如上图所示

File: en/docs/extended_topics/implement_data_loader.md [hunk 18]
```diff
@@ -0,0 +1,461 @@
...
+};
+```
+
+In addition to inheriting our `Dataset`'s `MiniDataset` class, OneFlow also build other `XXXDataset`, they can add additional features in the base of existing `Dataset`. For example, `RandomShuffleDataset` can be used to shuffle data, `BatchDataset` can be used to read batch data. When it is all done, we finally call `StartLoadThread`, which is used to start the loading thread. We will trigger the override function `MiniDataset::Next` in `StartLoadThread`. 
+
+The above construction of `MiniDataReader` can be used as a template. If you have no special requirements, you don't need to modify it in custom DataLoader. 
+
+## MiniDataset 
+
+For `MiniDataSet`, we only need to focus on the constructor and overridden `Next` function.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:238]: 同上，function 改为 method 较好（虽然有些资料会把method和function混用）。其它地方就自己检查下。

File: en/docs/extended_topics/implement_data_loader.md [hunk 19]
```diff
@@ -0,0 +1,461 @@
...
+
+The above construction of `MiniDataReader` can be used as a template. If you have no special requirements, you don't need to modify it in custom DataLoader. 
+
+## MiniDataset 
+
+For `MiniDataSet`, we only need to focus on the constructor and overridden `Next` function. 
+
+The constructor obtains user's settings through `Attr`. Then it will initialize the input stream according to the user's settings. In the following code, `JoinDirPath` is used to obtain all filenames according to the convention of dataset (the prefix, the amount of files, whether the filename number is complete, etc.). And `InitInStream` is to initialize the file in dataset as input stream (The member of `in_stream`), which is encapsulated by OneFlow, it will be used in `Next` function later.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:240]: whether the filename number is padded

File: en/docs/extended_topics/implement_data_loader.md [hunk 20]
```diff
@@ -0,0 +1,461 @@
...
+    auto pNums = sample_ptr->mut_data<double>();
+    pNums[0] = std::stod(numbers[0]);
+    pNums[1] = std::stod(numbers[1]);
+    ret.push_back(std::move(sample_ptr));
+
+    return ret;
+  }
+```
+
+In the above code, we call `in_stream_`'s `ReadLine` function to load file data into `string` object `sampleline`. Then we call`CommaSplit` to split the string by commas, convert to float, and place it to `TensorBuffer` object.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:279]: convert it to float value

File: en/docs/extended_topics/implement_data_loader.md [hunk 21]
```diff
@@ -0,0 +1,461 @@
...
+    pNums[1] = std::stod(numbers[1]);
+    ret.push_back(std::move(sample_ptr));
+
+    return ret;
+  }
+```
+
+In the above code, we call `in_stream_`'s `ReadLine` function to load file data into `string` object `sampleline`. Then we call`CommaSplit` to split the string by commas, convert to float, and place it to `TensorBuffer` object. 
+
+It is worth to mentioning that `_in_stream_` has two ways to read data from files:
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:281]: to mention

File: en/docs/extended_topics/implement_data_loader.md [hunk 22]
```diff
@@ -0,0 +1,461 @@
...
+int32_t PersistentInStream::ReadFully(char* s, size_t n);
+```
+
+`ReadLine` function read a row of file to `l` object; `ReadFully` function read `n` bytes data to the memory pointed by `s`. Both functions use 0 as return value on success. 
+
+`MiniDataset` complete the process of file to memory buffer. In next section, we will use `MiniParser` to set the buffer's content to Op's output. 
+
+## MiniParser
+
+`MiniPaser` inherits from `Parser`, we just need to override the `Parser` function.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:294]: override -> rewrite
override 指重载
这里是重写虚函数

File: en/docs/extended_topics/implement_data_loader.md [hunk 23]
```diff
@@ -0,0 +1,461 @@
...
+
+Notice that we use macro `MultiThreadLoop` in the procedure of setting `batch_data`'s data to the Op's output `dptr`. `MultiThreadLoop` allows our loop logic to be executed in multiple threads. It accept 2 parameters, the first parameter is the total number of loop; the second parameter is a callback function, its prototype is `void callback(size_t, i)`. OneFlow will create multiple threads, and call the callback function concurrently. The callback function's parameter `i` indicates the serial number of current loop, allowing us to divide data according to `i` and complete the business logic. 
+
+In the above code, we get the `i`th data in buffer through `batch_data->at(i).get()`, and set it to the location of the `i`th row in the output memory area. There are two columns in total. 
+
+## Data Decoder Operator
+
+Data Decoder operator is a normal operator. It accept the output of `DataReader` as its input, output one or multiple Blobs after some operations.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:324]: accept -> accepts
output one -> outputs one

File: en/docs/extended_topics/implement_data_loader.md [hunk 24]
```diff
@@ -0,0 +1,461 @@
...
+
+In the above code, we get the `i`th data in buffer through `batch_data->at(i).get()`, and set it to the location of the `i`th row in the output memory area. There are two columns in total. 
+
+## Data Decoder Operator
+
+Data Decoder operator is a normal operator. It accept the output of `DataReader` as its input, output one or multiple Blobs after some operations. 
+
+In  [ofrecord_decoder_ops.cpp](https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/user/ops/ofrecord_decoder_ops.cpp), we can see various decoders for OFRecord data. 
+
+The data processed by our Mini Dataloader is simple, so what MiniDecoder does is very simple. It just splits two columns data output from `DataReader` into two one-column outputs as `x` and `y`.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:328]: The data processed by our Mini Dataloader is simple, so the work done by MiniDecoder is also very simple.

File: en/docs/extended_topics/implement_data_loader.md [hunk 25]
```diff
@@ -0,0 +1,461 @@
...
+      *(out_dptr_x + i) = *(input+i*2);
+      *(out_dptr_y + i) = *(input+i*2 + 1);
+    });
+
+  }
+  //...
+};
+```
+
+We mainly get the input `in_blob` in `MiniDecoderKernel::Compute`, and then in multiple threads loop `MultiThreadLoop`, split the input data into `out_dptr_x` and `out_dptr_y`, they correspond to the output `x` and `y`.
```
Comment (doombeaker) [inline en/docs/extended_topics/implement_data_loader.md:380]: they correspond -> that correspond

**Review by doombeaker — APPROVED** [review]

**Pull request merged**